File: clients/os/src/apps/deployables/sources/repositories.ts
import type { Row } from "@znasllc-io/memql-sdk-core/client";

import { flatten } from "../../../kit/rows";

// The repositories a GitHub grant can see, as the picker reads them (epic
// memql#4915, design section A step 3).
//
// ===========================================================================
// A READING, NOT A FEED
// ===========================================================================
// `sourceRepositories` asks GitHub, through the cluster, what this person's
// installations hold right now. Nothing broadcasts a repository, and nothing
// could: the rows are not in this graph. So the picker prints WHEN it read
// and offers to look again, and this module has no subscription, no
// fingerprint and no arrival cue -- captioning liveness that is not there is
// the failure clients/os/README.md names for the on-demand surfaces.
//
// Everything here is pure -- a reply in, groups out -- so the picker's
// behaviour (what is grouped under whom, what a search matches, where a
// pending organization lands) is provable without rendering anything.

/** One repository the grant can reach. */
export interface RepositoryRow {
  /** `owner/name`, and the identity of the row: the picker keys on it. */
  fullName: string;
  owner: string;
  name: string;
  /** The clone URL the Source stop will carry. */
  url: string;
  private: boolean;
  /** GitHub's own word: `public`, `private`, `internal`. */
  visibility: string;
  defaultBranch: string;
  pushedAt: string;
  /** Which installation answered for it, so a later call can name one. */
  installationId: string;
}

/** One installation the grant reaches: an account or an organization. */
export interface InstallationRow {
  id: string;
  login: string;
  /** `User` or `Organization`, as GitHub spells them. */
  accountType: string;
  /** `all` or `selected`. */
  repositorySelection: string;
  suspended: boolean;
}

/** An organization whose installation an owner has not approved yet. */
export interface PendingInstallation {
  login: string;
}

/** One page of the answer. */
export interface RepositoryPage {
  repositories: RepositoryRow[];
  installations: InstallationRow[];
  pending: PendingInstallation[];
  /** The next page to ask for, or 0 when this was the last. */
  nextPage: number;
  /** `ok`, or the refusal code the read stopped on. */
  reason: string;
}

export const EMPTY_PAGE: RepositoryPage = {
  repositories: [],
  installations: [],
  pending: [],
  nextPage: 0,
  reason: "",
};

// ---------------------------------------------------------------------------
// Reading the reply
// ---------------------------------------------------------------------------

/**
 * A member list, whether it arrives as JSON text or as a decoded array.
 *
 * `packages/rows.ts` has the same two-shape reader for the analysis report,
 * and this is deliberately not that one promoted to the kit: `listOf` CASTS
 * its members, while every reader below PROJECTS -- each field named, each
 * default stated -- which is what keeps an added engine field from arriving
 * in this app unconsidered. Two different jobs that happen to start the same
 * way.
 */
function membersOf(row: Record<string, unknown>, key: string): unknown[] {
  const raw = row[key];
  if (Array.isArray(raw)) return raw;
  if (typeof raw === "string" && raw !== "") {
    try {
      const parsed: unknown = JSON.parse(raw);
      if (Array.isArray(parsed)) return parsed;
    } catch {
      // An unreadable list renders as no list. Never a thrown parse error:
      // one malformed member must not blank the whole picker, and an empty
      // picker already says "this connection reaches no repositories yet".
    }
  }
  return [];
}

function text(member: unknown, key: string): string {
  if (member === null || typeof member !== "object") return "";
  const v = (member as Record<string, unknown>)[key];
  return typeof v === "string" ? v : "";
}

/**
 * A boolean that may have crossed the wire as a word.
 *
 * The engine renders a scalar boolean as the STRING "true" on a builtin's
 * reply row (`packages/calls.ts`), and a member nested in a list arrives
 * decoded. Both spellings mean the same thing and anything else means false,
 * which is the reading that asserts least: `private` decides whether a chip
 * is drawn, and drawing none for a value this build cannot read is quieter
 * than claiming a repository is public.
 */
function flag(member: unknown, key: string): boolean {
  if (member === null || typeof member !== "object") return false;
  const v = (member as Record<string, unknown>)[key];
  return v === true || v === "true";
}

function count(row: Record<string, unknown>, key: string): number {
  const v = row[key];
  if (typeof v === "number" && Number.isFinite(v)) return Math.trunc(v);
  if (typeof v === "string" && v.trim() !== "") {
    const parsed = Number(v);
    if (Number.isFinite(parsed)) return Math.trunc(parsed);
  }
  return 0;
}

/**
 * Owner and name, from whichever of the three fields the reply carries.
 *
 * `fullName` is the identity, and the two halves are derived from it when
 * they are absent rather than left blank -- a group header reading nothing
 * and a row reading `acme/widget` would be one repository shown as two
 * mistakes. When the halves are present they WIN, because an owner that
 * contains a slash is a thing GitHub does not have but a parser should not
 * assume.
 */
function namesOf(member: unknown): { fullName: string; owner: string; name: string } {
  const fullName = text(member, "fullName").trim();
  let owner = text(member, "owner").trim();
  let name = text(member, "name").trim();
  if (owner === "" || name === "") {
    const cut = fullName.indexOf("/");
    if (cut > 0) {
      if (owner === "") owner = fullName.slice(0, cut);
      if (name === "") name = fullName.slice(cut + 1);
    }
  }
  return { fullName: fullName || (owner !== "" && name !== "" ? `${owner}/${name}` : ""), owner, name };
}

/**
 * The picker's reading of one `sourceRepositories` reply row.
 *
 * A member with no name at all is DROPPED: the picker keys on `fullName` and
 * a row with none could not be chosen, so rendering it would be a line that
 * does nothing when clicked.
 */
export function repositoryPageFrom(raw: Row | undefined | null): RepositoryPage {
  if (!raw) return { ...EMPTY_PAGE };
  const row = flatten(raw) as Record<string, unknown>;
  const repositories: RepositoryRow[] = [];
  for (const member of membersOf(row, "repositories")) {
    const { fullName, owner, name } = namesOf(member);
    if (fullName === "") continue;
    repositories.push({
      fullName,
      owner,
      name,
      url: text(member, "url"),
      private: flag(member, "private"),
      visibility: text(member, "visibility"),
      defaultBranch: text(member, "defaultBranch"),
      pushedAt: text(member, "pushedAt"),
      installationId: text(member, "installationId"),
    });
  }
  const installations: InstallationRow[] = [];
  for (const member of membersOf(row, "installations")) {
    const id = text(member, "id");
    if (id === "") continue;
    installations.push({
      id,
      login: text(member, "login"),
      accountType: text(member, "accountType"),
      repositorySelection: text(member, "repositorySelection"),
      suspended: flag(member, "suspended"),
    });
  }
  const pending: PendingInstallation[] = [];
  for (const member of membersOf(row, "pending")) {
    const login = text(member, "login").trim();
    // A pending entry with no login says only "something is waiting", which
    // names nobody to ask -- the entire content of this state.
    if (login === "") continue;
    pending.push({ login });
  }
  return {
    repositories,
    installations,
    pending,
    nextPage: count(row, "nextPage"),
    reason: text(row, "reason"),
  };
}

// ---------------------------------------------------------------------------
// Grouping, ordering and search
// ---------------------------------------------------------------------------

/** One owner's block in the picker: their repositories, or the sentence. */
export interface RepositoryGroup {
  owner: string;
  repositories: RepositoryRow[];
  /** Waiting for an owner of this organization to approve the app. */
  pending: boolean;
}

/** Whether a repository answers what somebody typed. Case-insensitive over
 *  the FULL name, so "acme/wid" and "WIDGET" both find `acme/widget`. */
export function matchesSearch(repo: Pick<RepositoryRow, "fullName">, search: string): boolean {
  const needle = search.trim().toLowerCase();
  if (needle === "") return true;
  return repo.fullName.toLowerCase().includes(needle);
}

function byName(a: string, b: string): number {
  return a.toLowerCase().localeCompare(b.toLowerCase());
}

/**
 * The picker's list: filtered, grouped by owner, ordered.
 *
 * PENDING GROUPS COME LAST, and that is the one ordering decision here. A
 * group with nothing in it is not a choice -- it is a note about a choice
 * that is not available yet -- and interleaving it alphabetically would put
 * a line nobody can act on in the middle of the ones they can. When the
 * approval lands the group moves up into the pickable half, which is exactly
 * the news somebody pressed Look again for.
 *
 * A pending organization is matched against the SEARCH by its login, so
 * typing the name of the organization you are waiting on finds the sentence
 * explaining why it has no repositories rather than hiding it.
 */
export function groupRepositories(
  page: Pick<RepositoryPage, "repositories" | "pending">,
  search: string,
): RepositoryGroup[] {
  const byOwner = new Map<string, RepositoryRow[]>();
  for (const repo of page.repositories) {
    if (!matchesSearch(repo, search)) continue;
    const held = byOwner.get(repo.owner);
    if (held) held.push(repo);
    else byOwner.set(repo.owner, [repo]);
  }

  const groups: RepositoryGroup[] = [...byOwner.entries()]
    .map(([owner, repositories]) => ({
      owner,
      repositories: repositories.slice().sort((a, b) => byName(a.name || a.fullName, b.name || b.fullName)),
      pending: false,
    }))
    .sort((a, b) => byName(a.owner, b.owner));

  const needle = search.trim().toLowerCase();
  const pending = page.pending
    .filter((p) => needle === "" || p.login.toLowerCase().includes(needle))
    // An organization that already answered with repositories is not
    // pending: the same login must never render twice, once as rows and
    // once as a sentence saying it has none.
    .filter((p) => !byOwner.has(p.login))
    .map((p) => ({ owner: p.login, repositories: [], pending: true }))
    .sort((a, b) => byName(a.owner, b.owner));

  return [...groups, ...pending];
}

/** How many repositories a set of groups holds, for the footer's count. */
export function repositoryCount(groups: readonly RepositoryGroup[]): number {
  return groups.reduce((n, g) => n + g.repositories.length, 0);
}

File: component/packages/repositories.go
package packages

import (
	"context"
	"errors"
	"strconv"
	"strings"

	memorynodes "github.com/znasllc-io/memql/component/database/memory-nodes"
	"github.com/znasllc-io/memql/component/packages/githubapp"
	"github.com/znasllc-io/memql/core/num"
)

// repositories.go -- the Source stop's repository picker (epic memql#4912, C7).
//
// A CAPABILITY RATHER THAN A QUERY, and the reason is where the answer lives:
// it is not in this cluster. What a grant can reach right now is what GitHub
// says, and GitHub's answer changes when somebody adds a repository to an
// installation with nothing to tell us. A row would be a cache nobody could
// invalidate, and a picker offering yesterday's list is the one failure a
// picker cannot survive.
//
// It runs under the CALLER's actor and resolves the CALLER's own grant. There
// is no argument by which one person lists another's repositories: an empty
// credentialId resolves through githubAppGrantForCaller, whose filter is the
// owner term and which takes no arguments at all, and a named one resolves
// through the same owner-scoped sealed read every other path uses.
//
// EVERY REFUSAL IS A `reason` IN THE REPLY rather than an error, because the
// picker renders in place: a person who has not connected GitHub sees Connect
// where the list would be, and a cluster with no app sees the token path. An
// error would collapse the stop into a failure banner and lose the repair.

// The reasons sourceRepositories answers with. Six, all but the first shared
// with a refusal code, so the OS keys one sentence per condition however it
// was reached.
const (
	RepositoriesReasonOK                = "ok"
	RepositoriesReasonNotConfigured     = CodeGithubAppNotConfigured
	RepositoriesReasonReconnectRequired = CodeReconnectRequired
	RepositoriesReasonCredentialMissing = CodeCredentialNotFound
	RepositoriesReasonCredentialRevoked = CodeCredentialRevoked
	RepositoriesReasonRateLimited       = ProbeReasonRateLimited
)

// SourceRepositoriesResult is the picker's whole answer.
type SourceRepositoriesResult struct {
	// Reason is exactly one of the RepositoriesReason* values. Everything
	// below is empty for anything but ok.
	Reason string `json:"reason"`
	// Repositories is what the grant reaches, across every installation, for
	// the requested page. Grouped by the client rather than here: `owner` is
	// carried on every row, and a server-side grouping would fix an order the
	// picker's search then has to undo.
	Repositories []PickerRepository `json:"repositories"`
	// Installations is every installation this grant can reach, so the picker
	// can say WHY a repository is missing -- an installation limited to
	// selected repositories, or suspended, is a different answer from one that
	// simply has none.
	Installations []PickerInstallation `json:"installations"`
	// Pending names the organizations whose installation is still waiting for
	// an owner's approval. By NAME, because the repair belongs to somebody
	// else and knowing whom to ask is the person's only useful next step.
	Pending []string `json:"pending"`
	// NextPage is the page to ask for next, or 0 when there is no more. Zero
	// rather than -1 or absent: the argument's own "empty or 0 means the first
	// page" makes 0 the value that cannot be mistaken for a page.
	NextPage int `json:"nextPage"`
}

// PickerRepository is one repository in the shape the Source stop needs: enough
// to group it, search it, show it and prefill a ref from it without a second
// call.
type PickerRepository struct {
	FullName       string `json:"fullName"`
	Owner          string `json:"owner"`
	Name           string `json:"name"`
	Url            string `json:"url"`
	Private        bool   `json:"private"`
	Visibility     string `json:"visibility"`
	DefaultBranch  string `json:"defaultBranch"`
	PushedAt       string `json:"pushedAt"`
	InstallationId string `json:"installationId"`
}

// PickerInstallation is one installation the grant reaches.
type PickerInstallation struct {
	Id                  string `json:"id"`
	Account             string `json:"account"`
	AccountType         string `json:"accountType"`
	RepositorySelection string `json:"repositorySelection"`
	// Suspended is an installation an owner has suspended. Carried rather
	// than filtered out: it explains why nothing under it works, and hiding
	// it would make those repositories simply absent.
	Suspended bool `json:"suspended"`
}

// SourceRepositories lists what the caller's GitHub App grant can reach.
func SourceRepositories(ctx context.Context, d *Deps, credentialId string, page int) (SourceRepositoriesResult, error) {
	res := SourceRepositoriesResult{
		Reason:        RepositoriesReasonOK,
		Repositories:  []PickerRepository{},
		Installations: []PickerInstallation{},
		Pending:       []string{},
	}
	if page < 1 {
		page = 1
	}
	if d.GitHubApp == nil || !d.GitHubApp.Configured() {
		// The operator's condition, answered before anything is resolved: a
		// cluster with no app has no grants either, and refusing here rather
		// than through a credential lookup is what lets the stop say "this
		// cluster offers the token path" instead of "no credential".
		res.Reason = RepositoriesReasonNotConfigured
		return res, nil
	}

	grant, reason, err := resolvePickerGrant(ctx, d, credentialId)
	if err != nil {
		return SourceRepositoriesResult{}, err
	}
	if reason != "" {
		res.Reason = reason
		return res, nil
	}

	installations, ierr := d.GitHubApp.UserInstallations(ctx, grant.Bearer)
	if ierr != nil {
		return pickerFailure(res, ierr)
	}

	ids := make([]string, 0, len(installations))
	for _, inst := range installations {
		res.Installations = append(res.Installations, PickerInstallation{
			Id:                  formatInstallationId(inst.Id),
			Account:             inst.Account.Login,
			AccountType:         inst.Account.Type,
			RepositorySelection: inst.RepositorySelection,
			Suspended:           strings.TrimSpace(inst.SuspendedAt) != "",
		})
		ids = append(ids, formatInstallationId(inst.Id))

		repos, total, rerr := d.GitHubApp.InstallationRepositories(ctx, grant.Bearer, inst.Id, page)
		if rerr != nil {
			return pickerFailure(res, rerr)
		}
		for _, repo := range repos {
			res.Repositories = append(res.Repositories, PickerRepository{
				FullName:       repo.FullName,
				Owner:          repo.Owner.Login,
				Name:           repo.Name,
				Url:            repo.HTMLURL,
				Private:        repo.Private,
				Visibility:     repo.Visibility,
				DefaultBranch:  repo.DefaultBranch,
				PushedAt:       repo.PushedAt,
				InstallationId: formatInstallationId(inst.Id),
			})
		}
		if total > page*githubapp.RepositoriesPerPage {
			res.NextPage = page + 1
		}
	}

	res.Pending = pendingInstallations(ctx, d, grant.Login)

	// THE GRANT'S STORED INSTALLATION IDS ARE REFRESHED FROM WHAT WE JUST
	// READ, and this is one of the three owner-actor paths that keep them
	// current (the connect callback and a probe under a grant are the others).
	// It is why this epic needs no privileged webhook automation: a delivery
	// names a GitHub identity rather than a MemQL user, so finding the grant
	// it belongs to would be a cross-owner read past the concept's own tier --
	// and nothing reads these ids on a hot path anyway, because the fetcher
	// asks GitHub which installation covers a repository, live. The list is a
	// display cache, and a display cache is exactly what an owner-actor
	// refresh is good enough for.
	//
	// BEST EFFORT: a failed refresh is a warning, never a failed listing. The
	// person asked for their repositories and they are in hand.
	if grant.Id != "" && d.Store != nil {
		if werr := d.Store.recordGrantInstallations(ctx, grant.Id, ids); werr != nil {
			d.log().Warn("packages: could not refresh a GitHub grant's installation ids",
				"component", "packages.repositories", "credential", grant.Id, "err", werr)
		}
	}
	return res, nil
}

// resolvePickerGrant finds the grant to list under, answering a typed reason
// rather than an error for everything a person can act on.
//
// A NAMED credential that is not a grant is refused as credential_not_found
// rather than listed under: a pasted token has no installations and no
// repository listing, so answering an empty list would say "you can reach
// nothing" about a credential that was never asked the question.
func resolvePickerGrant(ctx context.Context, d *Deps, credentialId string) (ResolvedCredential, string, error) {
	actor := actorFromContext(ctx)
	if d.PeekCredentials == nil {
		return ResolvedCredential{}, "", refuse(CodeSourceUnreadable,
			"this node cannot resolve credentials, so it cannot list a grant's repositories")
	}

	id := strings.TrimSpace(credentialId)
	if id == "" {
		if d.Store == nil {
			return ResolvedCredential{}, RepositoriesReasonCredentialMissing, nil
		}
		row, err := d.Store.githubAppGrantForCaller(ctx)
		if err != nil {
			return ResolvedCredential{}, "", err
		}
		if row == nil {
			// Not connected, or disconnected. The surface offers Connect.
			return ResolvedCredential{}, RepositoriesReasonCredentialMissing, nil
		}
		id = rowString(row, "id")
	}

	// The peek resolver, not the fetch one: listing repositories is a
	// question, and lastUsedAt is the record of a fetch (D11).
	grant, cerr := d.PeekCredentials(ctx, id, actor.UserId)
	if cerr != nil {
		switch RefusalCode(cerr) {
		case CodeCredentialNotFound:
			return ResolvedCredential{}, RepositoriesReasonCredentialMissing, nil
		case CodeCredentialRevoked:
			return ResolvedCredential{}, RepositoriesReasonCredentialRevoked, nil
		case CodeReconnectRequired:
			return ResolvedCredential{}, RepositoriesReasonReconnectRequired, nil
		case CodeGithubAppNotConfigured:
			return ResolvedCredential{}, RepositoriesReasonNotConfigured, nil
		}
		return ResolvedCredential{}, "", cerr
	}
	if !grant.IsGrant() {
		return ResolvedCredential{}, RepositoriesReasonCredentialMissing, nil
	}
	return grant, "", nil
}

// pickerFailure maps a GitHub failure onto the picker's typed reasons, keeping
// the shape of the reply intact so the stop renders in place.
func pickerFailure(res SourceRepositoriesResult, err error) (SourceRepositoriesResult, error) {
	empty := SourceRepositoriesResult{
		Repositories:  []PickerRepository{},
		Installations: []PickerInstallation{},
		Pending:       []string{},
	}
	switch {
	case errors.Is(err, githubapp.ErrReauthorize):
		empty.Reason = RepositoriesReasonReconnectRequired
		return empty, nil
	case errors.Is(err, githubapp.ErrNotConfigured):
		empty.Reason = RepositoriesReasonNotConfigured
		return empty, nil
	case githubapp.IsRateLimited(err):
		empty.Reason = RepositoriesReasonRateLimited
		return empty, nil
	}
	// Anything else is GitHub being unwell or unreachable, which is neither a
	// person's mistake nor a repair they can carry out. It stays an error, and
	// the stop shows the server's own sentence.
	return SourceRepositoriesResult{}, refuse(CodeSourceUnreadable,
		"this cluster could not read the repositories this GitHub connection reaches: %v", err)
}

// pendingInstallations names the organizations whose installation this person
// requested and an owner has not yet approved.
//
// GitHub surfaces these to the APP rather than to the person, so the call runs
// under the app JWT and the entries are matched on the REQUESTER's login --
// which is why the grant's own login is the argument. A FAILURE ANSWERS AN
// EMPTY LIST rather than failing the whole read: this endpoint is not
// available on every app plan, and a picker that refused to list anybody's
// repositories because it could not enumerate pending requests would trade the
// answer for the footnote.
func pendingInstallations(ctx context.Context, d *Deps, login string) []string {
	out := []string{}
	if d.GitHubApp == nil || strings.TrimSpace(login) == "" {
		return out
	}
	requests, err := d.GitHubApp.PendingInstallationRequests(ctx)
	if err != nil {
		return out
	}
	for _, r := range requests {
		if !strings.EqualFold(strings.TrimSpace(r.Requester.Login), strings.TrimSpace(login)) {
			continue
		}
		if account := strings.TrimSpace(r.Account.Login); account != "" {
			out = append(out, account)
		}
	}
	return out
}

// formatInstallationId renders an installation id as TEXT.
//
// Text on the wire and text on the row, because installationIds is a []string
// on the concept and a picker that answered a number would hand the client a
// value it has to re-render before it can compare the two.
func formatInstallationId(id int64) string {
	if id == 0 {
		return ""
	}
	return strconv.FormatInt(id, 10)
}

// ---------------------------------------------------------------------------
// The capability
// ---------------------------------------------------------------------------

func (i *Integration) handleSourceRepositories(ctx context.Context, args map[string]any, _ int) ([]memorynodes.MemoryNode, error) {
	deps, err := i.resolve()
	if err != nil {
		return nil, err
	}
	res, rerr := SourceRepositories(ctx, deps, stringArg(args, "credentialId"), intArg(args, "page"))
	if rerr != nil {
		return nil, rerr
	}
	return resultNode(map[string]any{
		"reason":        res.Reason,
		"repositories":  res.Repositories,
		"installations": res.Installations,
		"pending":       res.Pending,
		"nextPage":      res.NextPage,
	}), nil
}

// intArg reads a page number off the wire.
//
// Through core/num rather than a bare int(x): a decoded payload number arrives
// as float64 or int64, and Go leaves the out-of-range conversion
// implementation-defined -- on amd64 int(1e30) is hugely NEGATIVE, which would
// pass straight through the `< 1` guard as page one and hide the nonsense
// (memql#4779). OrZero is the named answer here because the caller already
// reads 0 as "the first page", so an unusable value lands on the behaviour an
// absent one has.
func intArg(args map[string]any, key string) int {
	if args == nil {
		return 0
	}
	switch v := args[key].(type) {
	case int:
		return v
	case int64:
		return num.Int64OrZero(v)
	case float64:
		return num.Float64OrZero(v)
	}
	return 0
}
